### PR TITLE
Add new Ransac methods for PnP based on USAC included in opencv since…

### PR DESCRIPTION
**Pull request by granger35**

### CODE REVIEW

**Review by jim-bcom — CHANGES_REQUESTED** [review]

File: src/SolARPoseEstimationSACPnpOpencv.cpp [hunk 1]
```diff
@@ -69,7 +76,10 @@ FrameworkReturnCode SolARPoseEstimationSACPnpOpencv::estimate(const std::vector<
     if (itr != convertPnPSACMethod.end())
         method = itr->second;
     else
+    {
+        LOG_WARNING("Pnp method called {} does not exist. Method ITERATIVE will be used instead", m_method);
```
[the remark this answers is not in the document]
Comment (duongnamduong) [reply src/SolARPoseEstimationSACPnpOpencv.cpp:80]: Resolved

File: src/SolARPoseEstimationSACPnpOpencv.cpp [hunk 2]
```diff
@@ -84,48 +94,90 @@ FrameworkReturnCode SolARPoseEstimationSACPnpOpencv::estimate(const std::vector<
...
+
+    if (m_reprojError >= 0)
+    {
+        std::vector<cv::Point2f> projected3D;
+        cv::projectPoints(worldCVPoints, raux, taux, m_camMatrix, m_camDistorsion, projected3D);
+        for (unsigned int i = 0; i < inliers_cv.rows; i++){
+            double err_reprj = norm(projected3D[inliers_cv.at<int>(i)] - imageCVPoints[inliers_cv.at<int>(i)]);
+            if (err_reprj < m_reprojError) {
+             inliers.push_back(inliers_cv.at<int>(i));
+             in2d.push_back(cv::Point2f(imagePoints[inliers_cv.at<int>(i)].getX(),imagePoints[inliers_cv.at<int>(i)].getY()));
```
[the remark this answers is not in the document]
Comment (duongnamduong) [reply src/SolARPoseEstimationSACPnpOpencv.cpp:131]: This code is removed

File: src/SolARPoseEstimationSACPnpOpencv.cpp [hunk 3]
```diff
@@ -84,48 +94,90 @@ FrameworkReturnCode SolARPoseEstimationSACPnpOpencv::estimate(const std::vector<
...
+             in3d.push_back(cv::Point3f(worldPoints[inliers_cv.at<int>(i)].getX(),worldPoints[inliers_cv.at<int>(i)].getY(),worldPoints[inliers_cv.at<int>(i)].getZ()));
+            }
+        }
+        /*
+        for (unsigned int i = 0; i < projected3D.size(); i++){
```
[the remark this answers is not in the document]
Comment (duongnamduong) [reply src/SolARPoseEstimationSACPnpOpencv.cpp:136]: Resolved

File: src/SolARPoseEstimationSACPnpOpencv.cpp [hunk 4]
```diff
@@ -84,48 +94,90 @@ FrameworkReturnCode SolARPoseEstimationSACPnpOpencv::estimate(const std::vector<
...
+    inliers.clear();
+
+    std::vector<cv::Point3f>in3d;
+    std::vector<cv::Point2f>in2d;
+
+    if (m_reprojError >= 0)
+    {
+        std::vector<cv::Point2f> projected3D;
+        cv::projectPoints(worldCVPoints, raux, taux, m_camMatrix, m_camDistorsion, projected3D);
+        for (unsigned int i = 0; i < inliers_cv.rows; i++){
```
[the remark this answers is not in the document]
Comment (duongnamduong) [reply src/SolARPoseEstimationSACPnpOpencv.cpp:127]: Resolved

File: src/SolARPoseEstimationSACPnpOpencv.cpp [hunk 5]
```diff
@@ -69,7 +76,10 @@ FrameworkReturnCode SolARPoseEstimationSACPnpOpencv::estimate(const std::vector<
     if (itr != convertPnPSACMethod.end())
         method = itr->second;
     else
+    {
+        LOG_WARNING("Pnp method called {} does not exist. Method ITERATIVE will be used instead", m_method);
         method = cv::SOLVEPNP_ITERATIVE;
```
[the remark this answers is not in the document]
Comment (duongnamduong) [reply src/SolARPoseEstimationSACPnpOpencv.cpp:81]: This code is removed

File: src/SolARPoseEstimationSACPnpOpencv.cpp [hunk 6]
```diff
@@ -84,48 +94,90 @@ FrameworkReturnCode SolARPoseEstimationSACPnpOpencv::estimate(const std::vector<
...
+    if (itr != convertPnPSACMethod.end())
+    {
+        postPNPMethod = itr->second;
+        // Test if method is adapted to non Ransac PNP (here non USAC methods)
+        if (postPNPMethod < cv::SOLVEPNP_ITERATIVE || postPNPMethod > cv::SOLVEPNP_IPPE_SQUARE)
+        {
+            postPNPMethod = -1;
+            LOG_WARNING("Post Pnp method called {} cannot be used, only non USAC mathods are possible. Post PNP will not applied ", m_postPNPMethod);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (duongnamduong) [reply src/SolARPoseEstimationSACPnpOpencv.cpp:171]: removed

File: src/SolARPoseEstimationSACPnpOpencv.cpp [hunk 7]
```diff
@@ -84,48 +94,90 @@ FrameworkReturnCode SolARPoseEstimationSACPnpOpencv::estimate(const std::vector<
...
+
+    // Applied a post PNP processing on Ransac PNP inliers.
+    if (postPNPMethod != -1)
+    {
+        cv::solvePnP(in3d, in2d, m_camMatrix, m_camDistorsion, raux,taux, true, postPNPMethod);
+    }
+    else {
+
```
[the remark this answers is not in the document]
Comment (duongnamduong) [reply src/SolARPoseEstimationSACPnpOpencv.cpp:179]: removed

File: tests/SolARTest_ModuleOpenCV_PnP/SolARTest_ModuleOpenCV_PnP_conf.xml [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        <configure component="SolARPoseEstimationSACPnpOpencv" name="UPNP">
+            <property name="method" type="string" value="UPNP"/>
+        </configure>
+        <configure component="SolARPoseEstimationSACPnpOpencv" name="IPPE">
+            <property name="method" type="string" value="IPPE"/>
+        </configure>
+        <configure component="SolARPoseEstimationSACPnpOpencv" name="USAC">
+            <property name="method" type="string" value="USAC"/>
+        </configure>
+        <configure component="SolARPoseEstimationSACPnpOpencv" name="USAC PARALLEL">
```
[the remark this answers is not in the document]
Comment (duongnamduong) [reply tests/SolARTest_ModuleOpenCV_PnP/SolARTest_ModuleOpenCV_PnP_conf.xml:45]: Resolved

File: tests/SolARTest_ModuleOpenCV_PnP/experiment_helpers.hpp [hunk 9]
```diff
@@ -0,0 +1,160 @@
+/******************************************************************************
```
[the remark this answers is not in the document]
Comment (duongnamduong) [reply tests/SolARTest_ModuleOpenCV_PnP/experiment_helpers.hpp:1]: removed

File: tests/SolARTest_ModuleOpenCV_PnP/main.cpp [hunk 10]
```diff
@@ -0,0 +1,285 @@
+#include <iostream>
```
[the remark this answers is not in the document]
Comment (duongnamduong) [reply tests/SolARTest_ModuleOpenCV_PnP/main.cpp:1]: Resolved

File: tests/SolARTest_ModuleOpenCV_PnP/main.cpp [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+
+//#include <boost/log/core.hpp>
+#include "xpcf/xpcf.h"
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <iomanip>
+#include <sstream>
+
+#include "random_generators.hpp"
```
[the remark this answers is not in the document]
Comment (duongnamduong) [reply tests/SolARTest_ModuleOpenCV_PnP/main.cpp:16]: Resolved

File: tests/SolARTest_ModuleOpenCV_PnP/main.cpp [hunk 12]
```diff
@@ -0,0 +1,285 @@
+#include <iostream>
+#include <fstream>
+
+#include <string>
+#include <vector>
+
+
+//#include <boost/log/core.hpp>
+#include "xpcf/xpcf.h"
```
[the remark this answers is not in the document]
Comment (duongnamduong) [reply tests/SolARTest_ModuleOpenCV_PnP/main.cpp:9]: Resolved

File: tests/SolARTest_ModuleOpenCV_PnP/main.cpp [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+using namespace Eigen;
+using namespace SolAR::PnPTest;
+
+#include "core/Log.h"
+
+// ADD COMPONENTS HEADERS HERE
+
+#include "xpcf/xpcf.h"
+
+#include "api/solver/pose/I3DTransformFinderFrom2D3D.h"
```
[the remark this answers is not in the document]
Comment (duongnamduong) [reply tests/SolARTest_ModuleOpenCV_PnP/main.cpp:31]: Resolved

File: src/SolARPoseEstimationSACPnpOpencv.cpp [hunk 14]
```diff
@@ -84,48 +94,90 @@ FrameworkReturnCode SolARPoseEstimationSACPnpOpencv::estimate(const std::vector<
...
+        {
+            postPNPMethod = -1;
+            LOG_WARNING("Post Pnp method called {} cannot be used, only non USAC mathods are possible. Post PNP will not applied ", m_postPNPMethod);
+        }
+    }
+
+    // Applied a post PNP processing on Ransac PNP inliers.
+    if (postPNPMethod != -1)
+    {
+        cv::solvePnP(in3d, in2d, m_camMatrix, m_camDistorsion, raux,taux, true, postPNPMethod);
```
[the remark this answers is not in the document]
Comment (duongnamduong) [reply src/SolARPoseEstimationSACPnpOpencv.cpp:176]: removed

File: src/SolARPoseEstimationSACPnpOpencv.cpp [hunk 15]
```diff
@@ -84,48 +94,90 @@ FrameworkReturnCode SolARPoseEstimationSACPnpOpencv::estimate(const std::vector<
...
+        return FrameworkReturnCode::_ERROR_;
+    }
+
+    int postPNPMethod = -1;
+    itr = convertPnPSACMethod.find(m_postPNPMethod);
+    if (itr != convertPnPSACMethod.end())
+    {
+        postPNPMethod = itr->second;
+        // Test if method is adapted to non Ransac PNP (here non USAC methods)
+        if (postPNPMethod < cv::SOLVEPNP_ITERATIVE || postPNPMethod > cv::SOLVEPNP_IPPE_SQUARE)
```
[the remark this answers is not in the document]
Comment (duongnamduong) [reply src/SolARPoseEstimationSACPnpOpencv.cpp:166]: removed

**Pull request merged**